Add integration tests and Github Actions

SteVwonder opened the pull request:
Add integration tests, github actions, docker setup, cleanup debug printing, and a few other various fixes.

Closes #1
Closes #2

SteVwonder committed:
move public code to lib.rs

SteVwonder committed:
fix compilation due to visibility

SteVwonder committed:
main: lookup valid KVS values for new Flux instance

Problem: "thing" and "other_thing" do not exist in Flux's KVS by
default, meaning that the default built binary always panics with
an OS error of FileNotFound.

Solution: Switch to looking up values in the KVS that exist within
a fresh Flux instance by default.

SteVwonder committed:
lib/main: add reactor_{run,stop} and use them in main

Problem: the main binary hangs indefinitely waiting for the reactor to
stop, and there are no safe interfaces to reactor_{run,stop}.

Solution: add safe interfaces for reactor run/stop and then use them in
the main function.  Also drop the `future.wait_for` since we are going
the async reactor route.

Note: unfortunately, it appears that not entering the reactor and just
going the declarative route does not work :(

SteVwonder committed:
fix warnings and implement suggestions from rustc

mostly consists of adding `dyn` to `Box<FnMut` reference and prefixing
unused variables with an underscore

SteVwonder committed:
src: apply `cargo fmt`

SteVwonder committed:
lib: fix UB when passing non-null URI to Flux::Open

Problem: CString.as_ptr()'s do not have lifetime information, so as soon
as their CString is destructed, they become invalid, and using them is
undefined behavior.

Solution: Follow the suggestion in the documentation for `as_ptr` to
bind the intermediate CString to a variable with a lifetime of the
function.

Reference: https://doc.rust-lang.org/std/ffi/struct.CString.html#method.as_ptr

SteVwonder committed:
tests: add SideFlux and convert main to integration tests

SteVwonder committed:
src: remove debug printing

SteVwonder committed:
docker: add dockerfile and associated run scripts

copied from flux-core and then modified (stripped down some)

SteVwonder committed:
github: add github action for PRs and merges to master

SteVwonder committed:
github: only run merge workflow on upstream not on forks

The docker hub secrets are only on the upstream

trws commented on Cargo.toml:
```diff
@@ -8,8 +8,8 @@ edition = "2018"
-git = "https://github.com/flux-framework/flux-sys-rs"
-branch = "master"
-
+git = "https://github.com/SteVwonder/flux-sys-rs"
+branch = "minor-tweaks-focal"
```
Do you mean to upstream this?

SteVwonder replied:
Just dropped this `fixup!` commit with the latest push. Good catch.

trws commented on src/errors.rs:
```diff
@@ -19,7 +19,7 @@ impl FluxResultCheck for std::os::raw::c_int {
         Self: std::marker::Sized,
     {
-        println!("{:?}", self);
+        //println!("{:?}", self);
```
Probably delete or put it under a condition rather than comment.

trws commented on src/lib.rs:
```diff
@@ -35,41 +35,52 @@ impl ToPtr for CString {
-            .flux_check()?;
+        let fut_ptr = unsafe {
+            flux_sys::flux_kvs_lookup(self.handle, std::ptr::null(), flags as i32, k.as_ptr())
+        }
+        .flux_check()?;
```
Is this a rustfmt artifact? If so it's fine it just looks really strange.

SteVwonder replied:
Yeah, it looks like it is.  I tried merging it in with the previous line then re-running the formatter and got the same thing. I agree it looks very weird as a singleton.  I like their new-line-per-method format when there are more chained methods, but for one it seems odd.

SteVwonder replied:
Oh! That reminds me that I should add cargo format to the CI.

trws replied:
That's a good idea.  I think it's bugging me because the thing on the previous line is a curly-brace, which is fine in rust but *very very odd* in C++.  I suppose it can come up, but man does it make my WTF sense tingle.  Either way going to resolve this, whether I like it in all places or not rustfmt is the way to go. 👍

trws commented on src/my_future.rs:
```diff
@@ -22,13 +18,13 @@ pub trait FromPtr: Sized {
 impl FromPtr for FluxFuture {
     fn from_ptr(f: *mut flux_sys::flux_future_t) -> FluxFuture {
-        eprintln!("new future: {:?}", f);
+        //eprintln!("new future: {:?}", f);
```
🤦 I'm sure I put these in, probably rip them out or put in a debug env var or something rather than commenting.

trws commented on src/my_future.rs:
```diff
@@ -63,26 +60,26 @@ pub trait MyFuture: FromPtr {
+                flux_sys::flux_future_continue(future.get_inner_mut(), next.get_inner_mut())
+                    .flux_check()
+                    .unwrap(); // panic on failure
+                               // next.forget(); // keep it from being reaped here
```
This is not a change, but seriously test this and see if that forget should be re-enabled.  Otherwise next is going to get reaped and I don't remember working through that it was always safe.  The refcount *should* handle it, but flux's refcount stuff on this was never meant to be automatable, and there are some odd corner cases.

trws commented on src/my_future.rs:
```diff
@@ -176,8 +175,9 @@ impl MyFuture for FluxFuture {
-        eprintln!("future being dropped {:?}", self);
+        //eprintln!("future being dropped {:?}", self);
         if !self.f.is_null() {
+            // TODO: how can we re-enable this?
```
This is related to the forget above, it's all about being able to track the lifetimes.  We can chat about this if you want, I'm not sure I can write out everything and be sure I capture it.  The trick is that the lifetime isn't really controlled by the system, it's dependent on what you do inside the callback, and automating interaction with it wasn't quite behaving when I first did it.  I think @garlick made a change to add a refcount  bump in the spot that was causing issues, but never got back to follow up on it.

SteVwonder replied:
>  We can chat about this if you want, I'm not sure I can write out everything and be sure I capture it.

Yeah, let's do that! I'm definitely interested in making this more bulletproof and stable.

trws reviewed (commented):
This generally looks really good.

SteVwonder committed:
github actions: add cargo fmt to CI

SteVwonder committed:
src: apply clippy's auto fixes

SteVwonder committed:
src: apply `cargo fmt` to the codebase

SteVwonder commented:
I believe I addressed all of the issues except the future refcounting.  @trws: would you be OK with merging this PR without fully addressing the refcounting on the futures, and instead we open an issue on that to address later?  I think it would be good to get the nightly CI up and running sooner rather than later, but I'll defer to your judgement.

Side-note: I think the CI isn't running b/c this PR is coming in from a fork, but I confirmed that the CI tests do pass by submitting a PR within my own fork: https://github.com/SteVwonder/flux-rs/pull/3

SteVwonder commented:
Oh. Since the last review, I also added rust format to the CI.  The rust formatter is a new, separate github action.

trws reviewed (approved):
LGTM.

Definitely agree on making the ref counting stuff a separate issue.  Hopefully it will turn out to be simple, but it seems like a really good idea to build that on top of this change set rather than trying to get it in here.

SteVwonder commented:
Thanks!

> Definitely agree on making the ref counting stuff a separate issue.

Ok.  Let's discuss further in #3.
